initial commit of new quickstart (#4333)
inmanta/inmanta-core: docs/quickstart.rst

Quickstart
***************

Inmanta is intended to manage complex infrastructures, often in the cloud or other virtualized environments.

> In this guide, we go for a less complex setup: Installing `containerlab <https://containerlab.dev/>`_ and configuring `SR Linux <https://learn.srlinux.dev/>`_ containers using **Inmanta service orchestrator** and ``gNMI``.

Reviewer: ```suggestion
In this guide we start simple and manage a 3 node CLOS network with spine and two leaf switches. In this guide we will install `containerlab <https://containerlab.dev/>`_ and configure `SR Linux <https://learn.srlinux.dev/>`_ containers using **Inmanta open source orchestrator** and ``gNMI``.
```

Author: `In this guide` is repeated here. it kind of sounds unnatural.  
Let's apply the suggestion here and then I'll rectify it later when everything is in order.

Reviewer: yeah, the two times in this guide is a mistake.

1. First, we use `Containerlab` to spin-up Inmanta server and its PostgreSQL database, then three `SR Linux` containers, connected in a CLOS like topology
2. After that, we configure IP addresses and OSPF on them using **Inmanta**.

.. note::

    This guide is meant to quickly set up an Inmanta LAB environment to experiment with.
    It is not recommended to run this setup in production, as it might lead to instabilities in the long term.

Prerequisites
----------------------------

``Docker``, ``Containerlab`` and ``Inmanta`` need to be installed on your machine and our ``SR Linux`` repository has to be cloned in order to proceed. Please make sure to follow the links below to that end.

1. `Install Docker <https://docs.docker.com/install/>`_.
2. `Install Containerlab <https://containerlab.dev/install/>`_.
3. Prepare a development environment by creating a `python virtual environment` and installing Inmanta:

   .. code-block:: sh

>        mkdir -p ~/.virtualenvs

Author: Do you mean installing in that folder is complex?  
I am following the same format as per the training and our videos.

       python3 -m venv ~/.virtualenvs/srlinux
       source ~/.virtualenvs/srlinux/bin/activate

>        pip install inmanta-core

Author: I believe we got the reason for installing `inmanta-core` on Slack yesterday?

4. Clone the `SR Linux examples <https://github.com/inmanta/examples/tree/master/Networking/SR%20Linux>`_ repository:

   .. code-block:: sh

       git clone https://github.com/inmanta/examples.git

> The cloned repository contains a **project.yml**, which looks like this:

Author: fixed in https://github.com/inmanta/inmanta-core/pull/4333/commits/44087c5057d06b4a27725222896be836f8edce50

    .. code-block:: yaml

        name: SR Linux Examples
        description: Provides examples for the SR Linux module
        author: Inmanta
        author_email: code@inmanta.com
        license: ASL 2.0
        copyright: 2022 Inmanta
        modulepath: libs
        downloadpath: libs
        repo:

>             - https://github.com/inmanta/

Reviewer: TODO: we need to fix this to provide a repo that contains the srlinux and yang modules.

        install_mode: release
        requires:

The ``modulepath`` setting defines that modules will be stored in ``libs`` directory.
The ``repo`` setting points to one or more Git repositories containing Inmanta modules.
The ``requires`` setting is used to pin versions of modules, otherwise the latest version is used.

5. Install the required modules:

   .. code-block:: sh

       cd Networking/SR\ Linux/
       inmanta project install

   .. note::

        should you face any errors at this stage, please contact us.

In the next sections we will showcase how to set up and configure ``SR Linux`` devices.

.. _lab:

Setting up the LAB
_________________________

Go to the `SR Linux` folder and then `containerlab` to spin-up the containers:

.. code-block:: sh

>     cd Networking/SR\ Linux/containerlab

Author: fixed in https://github.com/inmanta/inmanta-core/pull/4333/commits/44087c5057d06b4a27725222896be836f8edce50

    sudo clab deploy -t topology.yml

`Containerlab` will spin-up:

1. an `Inmanta` server
2. a `PostgreSQL` Database server
3. Three `SR Linux` network operating systems.

Depending on your system's horsepower, give them a few seconds/minutes to fully boot-up.

Connecting to the containers
______________________________

At this stage, you should be able to view the **Web Console** by navigating to:

> http://172.30.0.3:8888/dashboard

Reviewer: ```suggestion
http://172.30.0.3:8888/console
```

.. code-block:: sh

    cd ..

.. note::

    The rest of the this guide assumes commands are executed from the root path of the `SR Linux` folder, unless noted otherwise.

.. _inenv:

Create an Inmanta project and an environment
_____________________________________________

A project is a collection of related environments. (e.g. development, testing, production, qa,...). We need to have an environment to manage our infrastructure. An environment is a collection of resources, such as servers, switches, routers, etc.

There are two ways to create a project and an environment:

1. Using Inmanta CLI (**recommended**):
    .. code-block:: sh

        # Create a project called test
        inmanta-cli --host 172.30.0.3 project create -n test
        # Create an environment called SR_Linux
        inmanta-cli --host 172.30.0.3 environment create -p test -n SR_Linux --save

The first option, ``inmanta-cli``, will automatically create a ``.inmanta`` file that contains the required information about the server and environment ID. The compiler uses this file to find the server and to export to the right environment.

> 2. Using the Web Console: Connect to the Inmanta container http://172.30.0.3:8888/dashboard, click on the `Create new environment` button, provide a name for the project and the environment then click `submit`.

Reviewer: ```suggestion
2. Using the Web Console: Connect to the Inmanta container http://172.30.0.3:8888/console, click on the `Create new environment` button, provide a name for the project and the environment then click `submit`.
```

* Lines 68-74 instantiate an OSPF instance and OSPF ``Instance``, link to the ``OSPF instance``, provide a name, router ID, admin state and version.
* Lines 76-79 instantiate an ``OSPF`` object, link to the ``Protocols`` object and link to the ``OSPF instance``.
* Lines 81-84 instantiate an ``Area`` object, link to the ``OSPF instance`` and provide the area ID.
* Lines 86-89 instantiate an area ``Interface`` object, link to the ``OSPF area`` object and activates the OSPF on ``ethernet-1/1.0`` interface.

The rest of the configuration model follows the same method for ``leaf2`` and ``spine`` devices, with the only difference being the ``spine`` having two interfaces, subinterfaces and IP addresses and OSPF interface configuration.

Now, we can deploy the model by referring to `Deploy the configuration model`_ section.

Deploy the configuration model
____________________________________

To deploy the project, we must first register it with the management server by creating a project and an environment. We have covered this earlier at `Create an Inmanta project and an environment`_ section.

Export the ``interafces`` configuration model to the Inmanta server:

.. code-block:: sh

>     inmanta -vvv export interfaces.cf

Author: good catch. Fixing it.

>     inmanta -vvv export interfaces.cf

Author: fixed in https://github.com/inmanta/inmanta-core/pull/4333/commits/d7b8d8ec60e77f6b8114b60dd54b6d906534af6a

1. an `Inmanta` server
2. a `PostgreSQL` Database server
3. Three `SR Linux` network operating systems.

Depending on your system's horsepower, give them a few seconds/minutes to fully boot-up.

Connecting to the containers
______________________________

At this stage, you should be able to view the **Web Console** by navigating to:

http://172.30.0.3:8888/console

To get an interactive shell to the Inmanta server:

.. code-block:: sh

    docker exec -it clab-srlinux-inmanta-server /bin/bash

In order to connect to `SR Linux` containers, there are two options:

1. Using Docker:

.. code-block:: sh

    docker exec -it clab-srlinux-spine sr_cli
    # or
    docker exec -it clab-srlinux-leaf1 sr_cli
    # or
    docker exec -it clab-srlinux-leaf2 sr_cli

2. Using SSH (username and password is `admin`):

.. code-block:: sh

   ssh admin@clab-srlinux-spine
   ssh admin@clab-srlinux-leaf1
   ssh admin@clab-srlinux-leaf2

Then enter the `configuration mode` by typing:

.. code-block:: sh

    enter candidate

The output should look something like this:

.. code-block::

    Welcome to the srlinux CLI.
    Type 'help' (and press <ENTER>) if you need any help using this.

>     --{ running }--[  ]--

Reviewer 2: When trying the command in spine, I do not observe this output. Instead I get "--{ candidate shared default }--[  ]--" (without quotes)

Author: Could you verify from the console that the agent actually pushed the config to the spine?

Reviewer 3: But that this before the orchestrator actually does anything, isn't it?
 The issue reported here is I think that when you type `enter candidate` the cli doesn't show `--{ running }--[  ]--` but `--{ candidate shared default }--[ ]--` (which seems correct to me)

Author: I see, good catch!
